src/main/java/org/mastodon/feature/ui/FeatureComputationController.java

```diff
@@ -0,0 +1,141 @@
+package org.mastodon.feature.ui;
+
+import java.beans.PropertyChangeEvent;
+import java.beans.PropertyChangeListener;
+import java.text.DateFormat;
+import java.text.SimpleDateFormat;
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.Collection;
+import java.util.List;
+
+import javax.swing.JDialog;
+import javax.swing.JFrame;
+import javax.swing.JLabel;
+import javax.swing.JProgressBar;
+import javax.swing.SwingUtilities;
+
+import org.mastodon.feature.FeatureComputerService;
+import org.mastodon.feature.FeatureSpec;
+import org.mastodon.feature.FeatureSpecsService;
+import org.mastodon.graph.GraphChangeListener;
+import org.mastodon.revised.ui.util.EverythingDisablerAndReenabler;
+
+public class FeatureComputationController implements GraphChangeListener
+{
+
+	private static final DateFormat DATE_FORMAT = new SimpleDateFormat( "yyyy-MM-dd --- HH:mm:ss" );
+
+	private final JDialog dialog;
+
+	private final FeatureComputationPanel gui;
+
+	private final FeatureComputerService computerService;
+
+	private final FeatureComputationModel model;
+
+	private final PropertyChangeListener propertyChangeListener;
+
+	public FeatureComputationController( final FeatureSpecsService specsService, final FeatureComputerService computerService, final Collection< Class< ? > > targets )
+	{
+		this.computerService = computerService;
+		model = createModel( specsService, targets );
+		dialog = new JDialog( ( JFrame ) null, "Feature calculation" );
+		gui = new FeatureComputationPanel( model, targets );
+		dialog.getContentPane().add( gui );
+		dialog.pack();
+
+		gui.btnCompute.addActionListener( ( e ) -> compute() );
+		gui.btnCancel.addActionListener( ( e ) -> cancel() );
+
+		gui.progressBar.setString( "" );
+		propertyChangeListener = new PropertyChangeListener()
+		{
+			@Override
+			public void propertyChange( final PropertyChangeEvent evt )
+			{
+				if ( "progress".equals( evt.getPropertyName() ) )
```

Comment:
It seems that "progress" event is never sent anywhere.
What is the idea behind this?

Reply:
The `PropertyChangeListener` is an object that is passed to _feature computers_  as a dependency.
They are in charge of reporting `progress` if they want. An example is the gaussian mean intensity computer. Because it takes much much longer that all the other computers, it makes sense for it to report progress.